### ISSUE TITLE
Build flag to disable Lua and/or all C bits

### ISSUE THREAD

**Issue by tarleb** [jgm/pandoc#5164]
Prompted by hslua/hslua#73. Platform-independent support for hslua is sometimes tricky, and it can become an issue for porting efforts, e.g. hslua/hslua#67. Naturally, I'd rather see the underlying issues resolved, so this is only a stop-gap measure.

A flag to disable all C bits might, on the other hand, be helpful also to people trying to compile pandoc with GHCJS. That's much more involved, so disabling Lua would only be a first step.

**Comment by jgm** [comment]
See #4535 - there's a list there of all the cbits in our transitive dependencies.  It's hopeless to disable them all.

**Comment by jgm** [comment]
Perhaps it's worth considering a flag to disable lua support. However, I believe it's considered bad practice to have your API change depending on the settings of cabal flags.  Maybe it could be possible to keep the same API but raise an exception if lua support is not compiled in?

**Comment by tarleb** [comment]
I've played with this for a bit. The only way to make this work is to move all occurrences of the `Lua` type inside Text.Pandoc.Lua, to never expose a function with type `Lua` in its signature, and to stop exposing module Text.Pandoc.Lua. This is simple enough, but it seems that this would deprive library users of useful functionality. On the other hand, library users write Haskell instead of Lua to extend pandoc.

If the goal is to keep exposing "raw" pandoc Lua functionality, then one solution could be to leverage finer-grained library packages. E.g., having a *pandoc-core* package would make it possible to write a *pandoc-core-lua* and *pandoc-core-nolua* packages, which would then be used by pandoc proper. Cabal 2 supports having multiple libraries in the same cabal file, and finer grained modularity could solve a lot of portability issues. But there are obvious and undeniable costs in complexity and maintenance to this approach.

The seems to be of low priority now, as the hslua cross-compilation issue has been resolved. Should we close this for now in favor of #4535?
